### PR TITLE
Cargo, Crates, and Workspaces

### PR DESCRIPTION
**Pull request by listochkin**
- covers main Cargo features
- talks about crates ecosystem
- describes dependency resolution
- talks about workspaces
- ends with build performance and tooling around it

### CODE REVIEW

File: training-slides/src/using-cargo.md [hunk 1]
```diff
@@ -1 +1,417 @@
...
+* `--features`, `--all-features`, and `--no-default-features`
+* `--timings`
+
+## Putting it all together:
+
+`cargo +nightly run --target x86_64-apple-darwin --features "a b c dependency/feature" --timings`
+
+* use nightly Rust
+* enable features `a`, `b`, `c`, and a `feature` feature of a `dependency` crate
+* (assuming we use Apple Silicon computer) build a macOS executable for x86 processor and run it using build-in emulation (Rosetta2)
```
Comment (miguelraz) [inline training-slides/src/using-cargo.md:64]:
```suggestion
* (assuming we use an Apple Silicon computer) build a macOS executable for x86 processor and run it using built-in emulation (Rosetta2)
```

**Review by miguelraz — APPROVED** [review]
I am approving this, based on the understanding that it's an optional slide that only groups that really want to deep dive into the build mechanics will get when asking for it.
LGTM.

**Comment by jonathanpallant** [comment]
Ah, the image doesn't work in the downloaded artefacts. This is because the book is placed in `training-material-book`.

I think we need to patch mdslides to find all the images and copy them to the output-dir.

image

File: training-slides/src/using-cargo.md [hunk 2]
```diff
@@ -1 +1,417 @@
...
+
+## Timings Report
+
+![Cargo Build Report for Rust Analyzer](../book/images/rust-analyzer-cargo-build-timings.png)
+
+## Reading the report
+
+* Cargo can't start building a crate until all its dependencies have been built.
+    * Cargo only waits for `rustc` to produce an LLVM IR, further compilation by LLVM can run in background (purple)
+* A crate can't start building until its `build.rs` is build and finish running (yellow)
```
Comment (jonathanpallant) [inline training-slides/src/using-cargo.md:343]:
```suggestion
* A crate can't start building until its `build.rs` is built and finishes running (yellow)
```

**Comment by jonathanpallant** [comment]
Great slide deck - tons of very useful content. As discussed in the call, you could break the "timing" slides off into another deck, or stick a break slide in front of them so it's clear where can stop if you don't have time (ha ha) for that section.

I also made a couple of minor suggestions, and other comments above.

**Comment by jonathanpallant** [comment]
> I am approving this, based on the understanding that it's an optional slide that only groups that really want to deep dive into the build mechanics will get when asking for it.

I think all groups will at least need the introduction to workspaces, because the Connected DB uses a workspace.

File: training-slides/src/using-cargo.md [hunk 3]
```diff
@@ -1 +1,427 @@
 # Cargo Dependencies and Workspaces
+
+# Crates
+
+* Rust libraries are called `crates`
+* Crates are built using `cargo`
+* A crate can produce a single library and / or 1 or more executables
```
Comment (jonathanpallant) [inline training-slides/src/using-cargo.md:7]: *Technically*, a *package* can contain a single *library crate* and/or 1 more more *binary (i.e. executable) crates*

File: training-slides/src/using-cargo.md [hunk 4]
```diff
@@ -1 +1,427 @@
 # Cargo Dependencies and Workspaces
+
+# Crates
+
+* Rust libraries are called `crates`
```
Comment (jonathanpallant) [inline training-slides/src/using-cargo.md:5]: I don't know if I like the word library here, but I can't think of another word for "a lump of code".

Fun fact: IMHO crates.io is not a crate repository, it is a package repository - a package being the thing defined by the `Cargo.toml` file, which may contain crates.

File: training-slides/src/using-cargo.md [hunk 5]
```diff
@@ -1 +1,427 @@
...
+* Crates are built using `cargo`
+* A crate can produce a single library and / or 1 or more executables
+* A single crate can contain multiple modules
+* Multiple crates can be combined into a single Cargo workspace
```
Comment (jonathanpallant) [inline training-slides/src/using-cargo.md:9]:
```suggestion
* Multiple packages can be combined into a single Cargo workspace
```

File: training-slides/src/using-cargo.md (hunk 4 above)
Comment (jonathanpallant) [reply training-slides/src/using-cargo.md:5]: I think maybe:

* Rust code is arranged into packages
* A package can contain a single *library crate*, and/or 1 or more *binary crates*
* The package is described with a file called `Cargo.toml`
* The crates in that package are built using `cargo`

**Comment by jonathanpallant** [comment]
Slide 10 is now good, but the image still doesn't appear. I suggest we just hack the `build.sh` file to copy over all the image it finds in the source directory into the HTML output folder.

File: training-slides/src/using-cargo.md (hunk 4 above)
Comment (listochkin) [reply training-slides/src/using-cargo.md]: I went through all mentions of crates and packages and reworked all wording that could be lousy with the terms.

File: training-slides/src/using-cargo.md [hunk 6]
```diff
@@ -1,12 +1,13 @@
...
 
-* Rust libraries are called `crates`
-* Crates are built using `cargo`
-* A crate can produce a single library and / or 1 or more executables
-* A single crate can contain multiple modules
-* Multiple crates can be combined into a single Cargo workspace
+* Rust code is arranged into packages
+* A package is described by a `Cargo.toml` file
+* Building a package can produce a single library, and 0 or more executables
+    * these build artifacts are called "crates"
```
Comment (jonathanpallant) [inline training-slides/src/using-cargo.md:8]:
```suggestion
    * these are called *crates*
```

**Comment by jonathanpallant** [comment]
Rebased on main so the FFI stuff compiles (my bad for accidentally pushing to main).

**Pull request merged**